src/script/lua_api/l_server.cpp

```diff
@@ -498,29 +499,39 @@ int ModApiServer::l_notify_authentication_modified(lua_State *L)
 	return 0;
 }
 
-// get_last_run_mod()
-int ModApiServer::l_get_last_run_mod(lua_State *L)
+int ModApiServer::l_do_async_callback(lua_State *L)
 {
-	NO_MAP_LOCK_REQUIRED;
-	lua_rawgeti(L, LUA_REGISTRYINDEX, CUSTOM_RIDX_CURRENT_MOD_NAME);
-	std::string current_mod = readParam<std::string>(L, -1, "");
-	if (current_mod.empty()) {
-		lua_pop(L, 1);
-		lua_pushstring(L, getScriptApiBase(L)->getOrigin().c_str());
-	}
-	return 1;
-}
+	ServerScripting *script = getScriptApi<ServerScripting>(L);
 
-// set_last_run_mod(modname)
-int ModApiServer::l_set_last_run_mod(lua_State *L)
-{
-	NO_MAP_LOCK_REQUIRED;
-#ifdef SCRIPTAPI_DEBUG
-	const char *mod = lua_tostring(L, 1);
-	getScriptApiBase(L)->setOriginDirect(mod);
-	//printf(">>>> last mod set from Lua: %s\n", mod);
-#endif
-	return 0;
+	luaL_checktype(L, 1, LUA_TFUNCTION);
+	luaL_checktype(L, 2, LUA_TSTRING);
+	luaL_checktype(L, 3, LUA_TSTRING);
+
+	// Safely call string.dump on the function
+	lua_rawgeti(L, LUA_REGISTRYINDEX, CUSTOM_RIDX_GLOBALS_BACKUP);
+	if (!lua_isnil(L, -1))
+		lua_getfield(L, -1, "string");
+	else
+		lua_getglobal(L, "string");
+	lua_getfield(L, -1, "dump");
+	lua_pushvalue(L, 1);
```

Comment:
I don't think this is actually safe, afaik the `string` table is shared between the secure and insecure environment. To do this safely you need to store string.dump in a custom registry index at init or use `lua_dump` directly.

see:
https://github.com/minetest/minetest/blob/master/src/script/cpp_api/s_security.cpp#L88

and

https://github.com/minetest/minetest/blob/master/src/script/cpp_api/s_base.cpp#L100

Reply:
You are correct, this is problem even outside of this PR...

Reply:
Yeah mod security is a mess right now, not sure if anything can be done about it without breaking backwards compat.